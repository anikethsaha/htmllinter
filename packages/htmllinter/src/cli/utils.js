import { join } from 'path';
import chalk from 'chalk';

export const getDefaultConfigFilePath = () => {
  return join(process.cwd(), 'htmllinter.config.js');
};

export const getUserDefineConfigFilePath = (argv) => {
  let configFilePath = null;
  argv.forEach((arg, i) => {
    if (arg === '-c' || arg === '--config') {
      /**
       * arg[i + 1] is the value i.e path to config file.
       * Few things to check in order to get the exact file path name
       *
       * 1. it should be relative to `process.cwd()` -
       * 2. join(process.cwd(), arg[i+1],'htmllinter.config.js')
       *
       */
      const pathToConfig = argv[i + 1];

      configFilePath = join(process.cwd(), pathToConfig);
    }
  });

  return configFilePath === null ? undefined : configFilePath;
};

export const getConfig = (args) => {
  let configFilePath;
  /**
   * args[0] - input pattern
   * args[1..args.length-2] - we can find the "-c" | "--config" flags
   *
   * args[n] - if n is the position for flag name [property]
   * args[n+1] - will be the value
   *
   * So generally loop through the args array
   *
   */

  if (args.length < 2) {
    configFilePath = getDefaultConfigFilePath();
  } else {
    configFilePath =
      getUserDefineConfigFilePath(args) || getDefaultConfigFilePath();
  }

  let config;
  try {
    config = require(configFilePath);
  } catch (error) {
    process.stderr.write(
      `${chalk.yellow(
        error
      )}\n [HTMLLINTER] loading default config which is '{ }'`
    );
    return {};
  }

  return config;
};

export const help = chalk.yellowBright(`
Usage :

$ htmllinter <input> [Options]

input           - it is the input(html) file name, it accepts glob pattern.
                example :
                  - "input.html"
                  - "input"
                  - "/*.html"
                  - "."

Options
  -c, --config    - path to htmllinter.config.js.
                    default is "."
                    example :
                    - '.'
                    - './dir/'

Example

    $ htmllinter input.html
    $ htmllinter input -c ../
    $ htmllinter **/*
`);
