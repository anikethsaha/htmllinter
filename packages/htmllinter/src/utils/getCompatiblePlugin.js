/**
 * This function takes a plugin, and try to load that as if its compatbile of having
 * data pass through it. If its in-compatible, load the original plugins
 * example :
 *
 * Compatible of taking data (options) :  (optionsFromConfig = {}) => html => core.createHTMLLintPlugin(html, noEmptyTag, optionsFromConfig)
 *
 * Not Compatible of taking data (options) :  html => core.createHTMLLintPlugin(html, noEmptyTag, optionsFromConfig)
 *
 *
 * @param  {Function} tmpPlugin - the plugin which needs to be safely return
 * @returns {Function}
 */
export default (tmpPlugin) => {
  try {
    const clone = tmpPlugin.bind({});
    tmpPlugin = clone();
    // eslint-disable-next-line no-empty
  } catch (err) {}
  return tmpPlugin;
};
