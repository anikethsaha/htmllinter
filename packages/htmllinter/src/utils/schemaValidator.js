import Ajv from 'ajv';

const ajv = new Ajv({
  meta: false,
  useDefaults: true,
  validateSchema: false,
  missingRefs: 'ignore',
  verbose: true,
  schemaId: 'auto',
});

function getRuleOptionsSchema(schema) {
  if (!schema) {
    return null;
  }

  if (Array.isArray(schema)) {
    if (schema.length) {
      return {
        type: 'array',
        items: schema,
      };
    }
    return {
      type: 'array',
    };
  }

  return schema;
}

export const validateRuleSchema = (schema, option) => {
  let validate;
  try {
    const normalizedSchema = getRuleOptionsSchema(schema);

    validate = ajv.compile(normalizedSchema);
  } catch (error) {
    throw new Error(error.message || error);
  }

  validate(option);

  if (validate.errors) {
    throw new Error(
      validate.errors
        .map(
          (error) => `\tValue ${JSON.stringify(error.data)} ${error.message}.\n`
        )
        .join('')
    );
  }
};
