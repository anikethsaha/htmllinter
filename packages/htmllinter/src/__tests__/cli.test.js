import { resolve } from 'path';
import { run } from './utils/cliRunner';

describe('Testing CLI behavior', () => {
  describe('input pattern', () => {
    it('should show help when no arg is passed with proper err message', () => {
      const result = run(__dirname);
      expect(result.exitCode).toBe(0);
      expect(result.stderr).toMatchSnapshot(
        `[HTMLLINTER]: no argument passed. Please provide the input file(s) atleast`
      );

      expect(result.stdout.toString()).toMatchSnapshot('help');
    });

    it('should show help when when non-existing input is passed with proper err message', () => {
      const inputFileName = 'nonExist.html';
      const result = run(__dirname, [inputFileName]);
      expect(result.stderr).toMatchSnapshot(
        `[HTMLLINTER]: no file found with input pattern : ${inputFileName}`
      );
      expect(result.exitCode).toBe(0);
      expect(result.stdout.toString()).toMatchSnapshot('help');
    });

    it('should show message when wrong pattern is pass', () => {
      const pattern = 'dir,dir2';
      const result = run(__dirname, [pattern]);
      expect(result.stderr).toMatchSnapshot(
        `[HTMLLINTER]: no file found with input pattern : ${pattern}`
      );
      expect(result.exitCode).toBe(0);
    });
  });

  describe('Config flag', () => {
    it('should show message when correct path to config is passed but not present', () => {
      const configFilePath = './utils';
      const pattern = './fixtures/extendConfig/*.html';
      const result = run(__dirname, [pattern, '--config', configFilePath]);
      const errArr = result.stderr.split(' ');
      expect(errArr).toContain('Cannot');
      expect(errArr).toContain('find');
      expect(errArr).toContain('module');
      expect(result.stdout.toString()).toMatchSnapshot(
        'Output for loading default config { }'
      );
      expect(result.exitCode).toBe(0);
    });

    it('should show message when htmllinter.config.js is passed along with the path even if correct', () => {
      const configFilePath = './htmllinter.config.js';
      const pattern = '*.html';
      const result = run(resolve(__dirname, './fixtures/extendConfig'), [
        pattern,
        '--config',
        configFilePath,
      ]);

      const errArr = result.stderr.split(' ');
      expect(errArr).toContain('Cannot');
      expect(errArr).toContain('find');
      expect(errArr).toContain('module');
      expect(result.stdout.toString()).toMatchSnapshot(
        'Output for loading default config { }'
      );
      expect(result.exitCode).toBe(0);
    });
  });

  describe('default config : fixture : defaultConfig', () => {
    it('should not throw error, load default config (htmllinter.config.js is present) when no config arg is passed', () => {
      const pattern = 'input.html';
      const result = run(resolve(__dirname, './fixtures/defaultConfig'), [
        pattern,
      ]);

      expect(result.exitCode).toBe(1);
      expect(result.stdout.toString()).toMatchSnapshot('default config');
    });

    it('should not throw error, load default config (htmllinter.config.js is present) when no config arg is passed   : no-error', () => {
      const pattern = './fixtures/defaultConfig/noErr.html';
      const result = run(resolve(__dirname, './fixtures/defaultConfig'), [
        pattern,
      ]);

      expect(result.exitCode).toBe(0);
      expect(result.stdout.toString()).toMatchSnapshot(
        'default config no error'
      );
    });
  });

  describe('no config file : fixture : noConfig', () => {
    it('should load default config when neither config arg is passed nor htmllinter.config.js is present', () => {
      const pattern = 'input.html';
      const result = run(resolve(__dirname, './fixtures/noConfig'), [pattern]);
      const errArr = result.stderr.split(' ');
      expect(errArr).toContain('Cannot');
      expect(errArr).toContain('find');
      expect(errArr).toContain('module');
      expect(result.stdout.toString()).toMatchSnapshot(
        'Output for loading default config { }'
      );
      expect(result.exitCode).toBe(0);
    });
  });

  describe('standard config : fixture : extendConfig', () => {
    it('should show the linting errors  ( htmllinter.config.js present) ', () => {
      const configFilePath = '/';
      const pattern = 'input.html';
      const result = run(resolve(__dirname, './fixtures/extendConfig'), [
        pattern,
        '-c',
        configFilePath,
      ]);

      expect(result.exitCode).toBe(1);
      expect(result.stdout.toString()).toMatchSnapshot('standard config');
    });

    it('should not show the linting errors  ( htmllinter.config.js present) as non-Error files ', () => {
      const configFilePath = './';
      const pattern = 'noError.html';
      const result = run(resolve(__dirname, './fixtures/extendConfig'), [
        pattern,
        '-c',
        configFilePath,
      ]);

      expect(result.exitCode).toBe(0);
      expect(result.stdout.toString()).toMatchSnapshot('standard config');
    });
  });

  describe('standard config with rule off : fixture : extendConfigRuleOff', () => {
    it('should show the linting errors  ( htmllinter.config.js present) ', () => {
      const configFilePath = './fixtures/extendConfigRuleOff';
      const pattern = './fixtures/extendConfigRuleOff/input.html';
      const result = run(__dirname, [pattern, '-c', configFilePath]);
      expect(result.exitCode).toBe(1);
      expect(result.stdout.toString()).toMatchSnapshot(
        'standard config rule off'
      );
    });

    it('should not show the linting errors  ( htmllinter.config.js present) as non-Error files ', () => {
      const configFilePath = './';
      const pattern = 'noError.html';
      const result = run(resolve(__dirname, './fixtures/extendConfigRuleOff'), [
        pattern,
        '-c',
        configFilePath,
      ]);
      expect(result.exitCode).toBe(0);
      expect(result.stdout.toString()).toMatchSnapshot(
        'standard config rule off'
      );
    });
  });
});
