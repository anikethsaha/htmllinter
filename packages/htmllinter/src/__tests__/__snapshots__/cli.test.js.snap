// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CLI behavior Config flag should show message when correct path to config is passed but not present: Output for loading default config { } 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking ./fixtures/extendConfig/input.html [39m
[35m[39m[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking ./fixtures/extendConfig/noError.html [39m
[35m[39m

[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file ./fixtures/extendConfig/input.html[39m
[32m[39m
 [33mOutput for filename : ./fixtures/extendConfig/input.html [39m
[33m[39m



[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file ./fixtures/extendConfig/noError.html[39m
[32m[39m
 [33mOutput for filename : ./fixtures/extendConfig/noError.html [39m
[33m[39m
"
`;

exports[`Testing CLI behavior Config flag should show message when htmllinter.config.js is passed along with the path even if correct: Output for loading default config { } 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking input.html [39m
[35m[39m[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking noError.html [39m
[35m[39m

[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file input.html[39m
[32m[39m
 [33mOutput for filename : input.html [39m
[33m[39m



[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file noError.html[39m
[32m[39m
 [33mOutput for filename : noError.html [39m
[33m[39m
"
`;

exports[`Testing CLI behavior default config : fixture : defaultConfig should not throw error, load default config (htmllinter.config.js is present) when no config arg is passed   : no-error: default config no error 1`] = `
"[93m[39m
[93mUsage :[39m
[93m[39m
[93m$ htmllinter <input> [Options][39m
[93m[39m
[93minput           - it is the input(html) file name, it accepts glob pattern.[39m
[93m                example :[39m
[93m                  - \\"input.html\\"[39m
[93m                  - \\"input\\"[39m
[93m                  - \\"/*.html\\"[39m
[93m                  - \\".\\"[39m
[93m[39m
[93mOptions[39m
[93m  -c, --config    - path to htmllinter.config.js.[39m
[93m                    default is \\".\\"[39m
[93m                    example :[39m
[93m                    - '.'[39m
[93m                    - './dir/'[39m
[93m[39m
[93mExample[39m
[93m[39m
[93m    $ htmllinter input.html[39m
[93m    $ htmllinter input -c ../[39m
[93m    $ htmllinter **/*[39m
[93m[39m"
`;

exports[`Testing CLI behavior default config : fixture : defaultConfig should not throw error, load default config (htmllinter.config.js is present) when no config arg is passed: default config 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking input.html [39m
[35m[39m
 [33mOutput for filename : input.html [39m
[33m[39m
[90m0[39m  [31m7[39m:[31m5[39m  the tag < h3 > has no content.    [90mno-empty-tag[39m     [93merror[39m
[90m1[39m  [31m6[39m:[31m8[39m  duplicate ids 'asd adsasd' for a  [90mno-duplicate-id[39m  [93merror[39m"
`;

exports[`Testing CLI behavior input pattern should show help when no arg is passed with proper err message: [HTMLLINTER]: no argument passed. Please provide the input file(s) atleast 1`] = `"[41m[HTMLLINTER]:[49m [31mno argument passed. Please provide the input file(s) atleast[39m"`;

exports[`Testing CLI behavior input pattern should show help when no arg is passed with proper err message: help 1`] = `
"[93m[39m
[93mUsage :[39m
[93m[39m
[93m$ htmllinter <input> [Options][39m
[93m[39m
[93minput           - it is the input(html) file name, it accepts glob pattern.[39m
[93m                example :[39m
[93m                  - \\"input.html\\"[39m
[93m                  - \\"input\\"[39m
[93m                  - \\"/*.html\\"[39m
[93m                  - \\".\\"[39m
[93m[39m
[93mOptions[39m
[93m  -c, --config    - path to htmllinter.config.js.[39m
[93m                    default is \\".\\"[39m
[93m                    example :[39m
[93m                    - '.'[39m
[93m                    - './dir/'[39m
[93m[39m
[93mExample[39m
[93m[39m
[93m    $ htmllinter input.html[39m
[93m    $ htmllinter input -c ../[39m
[93m    $ htmllinter **/*[39m
[93m[39m"
`;

exports[`Testing CLI behavior input pattern should show help when when non-existing input is passed with proper err message: [HTMLLINTER]: no file found with input pattern : nonExist.html 1`] = `"[41m[HTMLLINTER]:[49m [31mno file found with input pattern : nonExist.html[39m"`;

exports[`Testing CLI behavior input pattern should show help when when non-existing input is passed with proper err message: help 1`] = `
"[93m[39m
[93mUsage :[39m
[93m[39m
[93m$ htmllinter <input> [Options][39m
[93m[39m
[93minput           - it is the input(html) file name, it accepts glob pattern.[39m
[93m                example :[39m
[93m                  - \\"input.html\\"[39m
[93m                  - \\"input\\"[39m
[93m                  - \\"/*.html\\"[39m
[93m                  - \\".\\"[39m
[93m[39m
[93mOptions[39m
[93m  -c, --config    - path to htmllinter.config.js.[39m
[93m                    default is \\".\\"[39m
[93m                    example :[39m
[93m                    - '.'[39m
[93m                    - './dir/'[39m
[93m[39m
[93mExample[39m
[93m[39m
[93m    $ htmllinter input.html[39m
[93m    $ htmllinter input -c ../[39m
[93m    $ htmllinter **/*[39m
[93m[39m"
`;

exports[`Testing CLI behavior input pattern should show message when wrong pattern is pass: [HTMLLINTER]: no file found with input pattern : dir,dir2 1`] = `"[41m[HTMLLINTER]:[49m [31mno file found with input pattern : dir,dir2[39m"`;

exports[`Testing CLI behavior no config file : fixture : noConfig should load default config when neither config arg is passed nor htmllinter.config.js is present: Output for loading default config { } 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking input.html [39m
[35m[39m

[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file input.html[39m
[32m[39m
 [33mOutput for filename : input.html [39m
[33m[39m
"
`;

exports[`Testing CLI behavior standard config : fixture : extendConfig should not show the linting errors  ( htmllinter.config.js present) as non-Error files : standard config 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking noError.html [39m
[35m[39m

[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file noError.html[39m
[32m[39m
 [33mOutput for filename : noError.html [39m
[33m[39m"
`;

exports[`Testing CLI behavior standard config : fixture : extendConfig should show the linting errors  ( htmllinter.config.js present) : standard config 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking input.html [39m
[35m[39m
 [33mOutput for filename : input.html [39m
[33m[39m
[90m0[39m  [31m7[39m:[31m5[39m  the tag < p > has no content.                                          [90mno-empty-tag[39m                  [93merror[39m
[90m1[39m  [31m8[39m:[31m5[39m  the tag < h3 > has no content.                                         [90mno-empty-tag[39m                  [93merror[39m
[90m2[39m  [31m6[39m:[31m8[39m  duplicate ids 'asd adsasd' for a                                       [90mno-duplicate-id[39m               [93merror[39m
[90m3[39m  [31m7[39m:[31m5[39m  the attribute \\"enable\\" boolean value \\"true\\", use \\"<p enable>\\" instead  [90mno-bool-true-explicit-define[39m  [93merror[39m"
`;

exports[`Testing CLI behavior standard config with rule off : fixture : extendConfigRuleOff should not show the linting errors  ( htmllinter.config.js present) as non-Error files : standard config rule off 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking noError.html [39m
[35m[39m

[42m[1m[HTMLLINTER] :[22m[49m [32m0 errors found for file noError.html[39m
[32m[39m
 [33mOutput for filename : noError.html [39m
[33m[39m
"
`;

exports[`Testing CLI behavior standard config with rule off : fixture : extendConfigRuleOff should show the linting errors  ( htmllinter.config.js present) : standard config rule off 1`] = `
"[42m[37m[1m[HTMLLINTER] :[22m[39m[49m[35m Checking ./fixtures/extendConfigRuleOff/input.html [39m
[35m[39m
 [33mOutput for filename : ./fixtures/extendConfigRuleOff/input.html [39m
[33m[39m
[90m0[39m  [31m6[39m:[31m8[39m  duplicate ids 'asd adsasd' for a  [90mno-duplicate-id[39m  [93merror[39m"
`;
